src/Nethermind/Nethermind.Merge.Plugin/GC/GCKeeper.cs

```diff
@@ -0,0 +1,98 @@
+// SPDX-License-Identifier: LGPL-3.0-only
+
+using System;
+using System.Runtime;
+using System.Threading.Tasks;
+using FastEnumUtility;
+using Nethermind.Core.Extensions;
+using Nethermind.Logging;
+
+namespace Nethermind.Merge.Plugin.GC;
+
+public class GCKeeper
+{
+    private readonly IGCStrategy _igcStrategy;
+    private readonly ILogger _logger;
+    private static readonly long _defaultSize = 512.MB();
+    private Task _gcScheduleTask = Task.CompletedTask;
+
+    public GCKeeper(IGCStrategy igcStrategy, ILogManager logManager)
+    {
+        _igcStrategy = igcStrategy;
+        _logger = logManager.GetClassLogger<GCKeeper>();
+    }
+
+    public IDisposable TryStartNoGCRegion(long? size = null)
+    {
+        size ??= _defaultSize;
+        if (_igcStrategy.ShouldTryToPreventGCDuringBlockProcessing())
+        {
+            bool noGcRegion = System.GC.TryStartNoGCRegion(size.Value, true);
+            return new NoGCRegion(this, noGcRegion ? FailCause.None : FailCause.GC, size, _logger);
+        }
+
+        return new NoGCRegion(this, FailCause.Strategy, size, _logger);
+    }
+
+    private enum FailCause
+    {
+        None,
+        Strategy,
+        GC,
+    }
+
+    private class NoGCRegion : IDisposable
+    {
+        private readonly GCKeeper _gcKeeper;
+        private readonly FailCause _failCause;
+        private readonly long? _size;
+        private readonly ILogger _logger;
+
+        internal NoGCRegion(GCKeeper gcKeeper, FailCause failCause, long? size, ILogger logger)
+        {
+            _gcKeeper = gcKeeper;
+            _failCause = failCause;
+            _size = size;
+            _logger = logger;
+        }
+
+        public void Dispose()
+        {
+            if (_failCause == FailCause.None)
+            {
+                if (GCSettings.LatencyMode == GCLatencyMode.NoGCRegion)
+                {
+                    try
+                    {
+                        System.GC.EndNoGCRegion();
+                    }
+                    catch (InvalidOperationException)
+                    {
+                        if (_logger.IsDebug) _logger.Debug($"Failed to keep in NoGCRegion with Exception with {_size} bytes");
+                    }
+                }
+                else if (_logger.IsDebug) _logger.Debug($"Failed to keep in NoGCRegion with {_size} bytes");
+            }
+            else if (_logger.IsDebug) _logger.Debug($"Failed to start NoGCRegion with {_size} bytes with cause {_failCause.FastToString()}");
+
+            _gcKeeper.ScheduleGC();
+        }
+    }
+
+    private void ScheduleGC()
+    {
+        if (_gcScheduleTask.IsCompleted) _gcScheduleTask = ScheduleGCInternal();
+    }
+
+    private async Task ScheduleGCInternal()
+    {
+        int gcToCollect = Math.Min(_igcStrategy.ShouldForceGCBetweenBlockProcessing(), System.GC.MaxGeneration);
+        if (gcToCollect >= 0)
+        {
+            await Task.Delay(1000);
```

Comment:
Why is this delay here?

Reply:
good question, I want all the responses of engine API be processed before invoking GC. Normally we should get block every 12s, so lets say we process one in 2s, then delay 1s, then invoke GC.

Reply:
I'd leave it as a comment to not introduce a magic number 😇

Reply:
hmm, what if we have a block that will take a long time to process with a short slot time (5s for example) and we have a block that will take 7s for processing? Just thinking out loud